### ISSUE TITLE
Figure out why Nemo-Serve crashes if given too much text

### ISSUE THREAD

**Issue by gaurav** [RENCI-NER/nemo-serve#19]
We discovered previously that if Nemo-Serve was given more than a certain amount of text (somewhere around 100 characters, I think?) it would crash on the GPU, requiring Nemo-Serve to be restarted. This issue covers trying to diagnose why this is happening, or at least collecting enough information for us to submit it as an issue to https://github.com/NVIDIA/NeMo.

**Comment by gaurav** [comment]
@YaphetKG I think you told me once that re-setting up the Nemo-Jupyter instance we once had might help with this, since we could replicate the bug consistently in that. Is that right? If so, could you please either start that instance back up or point us at the Helm file needed to start it on Sterling? Thanks!

**Comment by YaphetKG** [comment]
Hi @gaurav , @mbacon-renci , its running at (https://github.com/RENCI-NER/Helm-charts/blob/main/nemo-helm/values.yaml#L80) , (https://nemo-jupyter.apps.renci.org/tree/notebooks) , its installed via that chart in the ner namespace , using the default value file there. The release name is nemo-train in helm. 

the jupyter path `tree/notebooks` is where you'd want to store your notebooks as its backed by a volume , saving your notebook else where might be lost on next restart.

**Comment by mbacon-renci** [comment]
I started looking at this last night and am getting a 404 on that nemo-juptyer.apps.renci.org URL, and I’ve exhausted my own attempts to debug it. Is that something that I should be able to get to, or do I need to launch a version of it in my own namespace?

Thanks,
Michael

From: YaphetKG ***@***.***>
Date: Friday, October 7, 2022 at 9:02 AM
To: RENCI-NER/nemo-serve ***@***.***>
Cc: Bacon, Michael T ***@***.***>, Mention ***@***.***>
Subject: Re: [RENCI-NER/nemo-serve] Figure out why Nemo-Serve crashes if given too much text (Issue #19)

Hi @gaurav<https://github.com/gaurav> , @mbacon-renci<https://github.com/mbacon-renci> , its running at (https://github.com/RENCI-NER/Helm-charts/blob/main/nemo-helm/values.yaml#L80) , (https://nemo-jupyter.apps.renci.org/tree/notebooks) , its installed via that chart in the ner namespace , using the default value file there. The release name is nemo-train in helm.

the jupyter path tree/notebooks is where you'd want to store your notebooks as its backed by a volume , saving your notebook else where might be lost on next restart.

**Comment by mbacon-renci** [comment]
After forking Yaphet's notebook, I parameterized the window size setting to higher level calls for easier testing, and wrote timings, parameters, and results to output files to try to chase this down. After trying both some randomized trials and some sequenced ones, here's what I've found so far:

**tl;dr, it's probably just the library hitting a limit of GPU memory, and the best proposed solution I've found is something like the sliding window solution. I now intend to investigate the effect of the size of the sliding window on model output quality and performance**

#### CUDA errors

As originally written, the most common failure mode was that something like the following error would be given before processing completed: `CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)`

By trying a range of window sizes, I attempted to suss out whether this was likely either an algorithmic or misjudged integer type error, which might see the error occur at a specific cutoff point (I was suspicious of 256 or related numbers, which could mean that a counter or an index was going beyond an 8-bit integer or some such). Instead, there was no consistent point that consistently demarcated successful model runs from erroring ones. Errors occurred with window sizes as low as 250, and occasionally the model would successfully run with a window size of 350 or 400.

Based on online reports, the error given appears to be indicative of the machine running out of GPU memory (_e.g._, https://discuss.pytorch.org/t/cuda-error-cublas-status-not-initialized-when-calling-cublascreate-handle/125450/8). Multiple sites suggested reducing batch size to avoid this, which is effectively what the sliding window batching function does already.

Following  online suggestions, I have found setting the CUDA_LAUNCH_BLOCKING environment variable to 1 changes the error message from the above error to `CUDA error: device-side assert triggered`. No change in the output, success rate, or behavior of the model was observed with this variable set.

#### Jupyter kernel crashes

Sometimes the model fails with a different failure mode, in which before the model complete, the jupyter UI raises an error that the jupyter kernel has crashed, and that it will restart it. This mode occurred with large and small windows, and while its lower frequency makes it harder to speak definitively, it appears to occur frequently when the model is re-initialized and re-run immediately after hitting the CUDA error above.

It is my supposition that this may be from a failed model insufficiently releasing GPU resources, and subsequent calls on the GPU being unexpectedly blocked, causing likely a bus error and a python kernel crash.

When running model runs re-using the same model instance and not re-initializing each time, this failure mode never occurred.

#### What next?

In doing minor instrumentation on the notebook-based model test, I kept a fair amount of detail from successive model runs. I want to look at these runs and see if how model output quality and runtime changes based on window size.

**Comment by mbacon-renci** [comment]
Migrated to Jira: https://renci.atlassian.net/browse/DUG-72?atlOrigin=eyJpIjoiMjU2NmVjYWNlYjhlNDM5N2FlZWE0ZjE5ZjdjMTY2YzAiLCJwIjoiaiJ9
